module/zfs/txg.c

```diff
@@ -549,8 +550,20 @@ txg_sync_thread(void *arg)
 			txg_thread_wait(tx, &cpr, &tx->tx_quiesce_done_cv, 0);
 		}
 
-		if (tx->tx_exiting)
-			txg_thread_exit(tx, &cpr, &tx->tx_sync_thread);
+		if (tx->tx_exiting) {
+			if (checked_quiescing) {
+				txg_thread_exit(tx, &cpr, &tx->tx_sync_thread);
+			} else {
+				while (tx->tx_threads != 1)
+					txg_thread_wait(tx, &cpr,
+					    &tx->tx_exit_cv, 0);
+				if (tx->tx_quiesced_txg)
+					checked_quiescing = B_TRUE;
+				else
+					txg_thread_exit(tx, &cpr,
+					    &tx->tx_sync_thread);
+			}
```

Comment:
This feels a little overly complicated.  It should be sufficient to always wait for `tx->tx_threads == 1`, and then only when `if (tx->tx_quiesced_txg != 0)` do we know there's one last txg to sync.  Maybe something like this instead (untested)?  Or is there a case this doesn't cover?

```c
                /*
                 * Wait until the quiesce thread has exited to ensure every
                 * quiesced txg has been synced before exiting.
                 */
                if (tx->tx_exiting) {
                        while (tx->tx_threads != 1)
                                txg_thread_wait(tx, &cpr, &tx->tx_exit_cv, 0);

                        if (tx->tx_quiesced_txg == 0)
                                txg_thread_exit(tx, &cpr, &tx->tx_sync_thread);
                }
```

Reply:
@behlendorf thanks for the review! I agree the extra complexity is completely unnecessary. I have made edits as you suggested.